### ISSUE TITLE
How to freeze some params when training a GPT model?

### ISSUE THREAD

**Issue by lvcc2018** [NVIDIA/Megatron-LM#345]
I tried to set params.required_grad to False when initializing the model, but some error in the optimizer occurred. So how to freeze some params correctly?

**Comment by lwmlyy** [comment]
+1

**Comment by Wenjun-Peng** [comment]
+1

**Comment by jubgjf** [comment]
I made changes in three places:
1. Add `torch.no_grad()` in `TransformerLanguageModel.forward` when calling the encoder:
```python
# Run encoder.
args = get_args()
if args.freeze_weight:  # add `--freeze-weight` option in arguments.py
    with torch.no_grad():
        if enc_hidden_states is None:
            if self.encoder is not None:
                encoder_output = self.encoder(
                    encoder_input,
                    enc_attn_mask,
                    inference_params=inference_params)
            else:
                encoder_output = self.encoder_hidden_state
        else:
            encoder_output = enc_hidden_states.to(encoder_input.dtype)
else:
    # remain unchanged from the original code
    if enc_hidden_states is None:
        if self.encoder is not None:
            encoder_output = self.encoder(
                encoder_input,
                enc_attn_mask,
                inference_params=inference_params)
        else:
            encoder_output = self.encoder_hidden_state
    else:
        encoder_output = enc_hidden_states.to(encoder_input.dtype)
```
2. Prevent layernorm and rmsnorm from entering train mode by overriding `train` method of `MixedFusedLayerRMSNorm` and `MixedFusedLayerNorm`:
```python
def train(self, mode):
    args = get_args()
    if args.freeze_weight:
        super().train(False)
        if mode:
            print_rank_0(
                f"Attempt to set RMSNorm to train mode, but we set freeze_weight == True, so RMSNorm is still in eval mode")
        assert not self.training
    else:
        super().train(mode)
```
3. Recursively set `required_grad = False` when loading state dict:
```python
# in `TransformerLanguageModel.load_state_dict`

# Encoder.
if self.add_encoder:
    ...
    self.encoder.load_state_dict(state_dict_, strict=strict)
    args = get_args()
    if args.freeze_weight:
        self.no_requires_grad(self.encoder)
```
```python
def no_requires_grad(self, module):
    if len(module._modules.keys()) != 0:
       for submodule_name, submodule in module._modules.items():
            print(f"Enter {submodule_name}")
            self.no_requires_grad(submodule)
    else:
        print(f"Setting {module._get_name()}.requires_grad = False")
        module.requires_grad = False
        print(f"Leave")
```

I freeze all the parameters in `no_requires_grad`, except embedding and lm_head (output_dense), we can also selectively freeze the specified weight by judging the `submodule_name`.

I think wrapping `torch.no_grad` and setting `require_grads = True` are the same, but I haven't tried enabling only one of them.

If we only need to freeze partial parameters in one layer, we can manually clear gradients in a tensor after `loss.backward` and before `optimizer.step`. In the code below I extended tokenizer vocab_size, I want to freeze old embeddings&lm_head and only train new embeddings&lm_head, so I set their grad to 0 by `xxx.main_grad[<some range>] = 0`.
```python
# in `train_step`

losses_reduced = forward_backward_func(
    forward_step_func, data_iterator, model,
    optimizer, fwd_bwd_timers, forward_only=False)
timers('forward-backward').stop()

# Empty unused memory.
if args.empty_unused_memory_level >= 1:
    torch.cuda.empty_cache()

# Reduce gradients.
optimizer.reduce_model_grads(args, timers)

# Vision gradients.
...

# Clean gradients of old embeddings if we are extending vocab
if args.freeze_weight:
    assert args.extend_vocab and args.old_vocab_size != -1
    unwrapped_model = unwrap_model(model[0],
                                   (torchDDP, LocalDDP, Float16Module))
    embeddings = unwrapped_model._modules["language_model"]._modules["embedding"]._modules["word_embeddings"]
    lm_head = unwrapped_model._modules["language_model"]._modules["output_dense"]

    new_vocab_start_index = embeddings.vocab_start_index
    old_vocab_end_index = args.old_vocab_size - 1

    if old_vocab_end_index >= new_vocab_start_index:
        embeddings._parameters['weight'].main_grad[:old_vocab_end_index - new_vocab_start_index + 1] = 0
        lm_head._parameters['weight'].main_grad[:old_vocab_end_index - new_vocab_start_index + 1] = 0

    # Update parameters.
    timers('optimizer', log_level=1).start(barrier=args.barrier_with_L1_time)
    update_successful, grad_norm, num_zeros_in_grad = optimizer.step(args, timers)
    timers('optimizer').stop()
```

If `tensor-model-parallel-size` > 1, each layer of the loaded model is already split, so we need to pay special attention to the range to clear gradient.

**Comment by ygfrancois** [comment]
I want to warm-up train only the word embedding and lm head.
I use pipeline parallel size == 4, and freeze encoder. That means there are no weights to train in two nodes, that causes the optimizer bug on these two nodes. 

Do you have any ideas about this problem?

**Comment by jubgjf** [comment]
@ygfrancois Can you show the error message?

In addition, I think tp=4 will distribute embedding and lmhead equally to each gpu. For example, when vocab size is 32000, theoretically the embedding size on each gpu is 32000/4=8000. Since alignment is 128 (controlled by `--make_vocab_size_divisible_by`), The actual embedding size will be 8064. That means each gpu should have trainable parameters.

**Comment by ygfrancois** [comment]
> @ygfrancois Can you show the error message?
> 
> In addition, I think tp=4 will distribute embedding and lmhead equally to each gpu. For example, when vocab size is 32000, theoretically the embedding size on each gpu is 32000/4=8000. Since alignment is 128 (controlled by `--make_vocab_size_divisible_by`), The actual embedding size will be 8064. That means each gpu should have trainable parameters.

thank you for your reply.
tp=4 is ok for me. Here the problem is that i use pp=4 and tp=8 to train 70B model.
pp=4 causes the problem. On node0 and node3, the initial is ok because there are word embedding and lm_head to train on each GPU. But on node1 and node2, there are only encoder params which are all freezed.

log:

`

[default3]:  File "/root/code/M6-BoYan/finetune_online.py", line 117, in <module>
[default3]:    pretrain(
[default3]:  File "/root/code/M6-BoYan/megatron/training.py", line 110, in pretrain
[default3]:    model, optimizer, opt_param_scheduler = setup_model_and_optimizer(
[default3]:  File "/root/code/M6-BoYan/megatron/training.py", line 428, in setup_model_and_optimizer
[default3]:    optimizer = get_megatron_optimizer(model, no_wd_decay_cond,
[default3]:  File "/root/code/M6-BoYan/megatron/optimizer/__init__.py", line 78, in get_megatron_optimizer
[default3]:    optimizer = Adam(param_groups,
[default3]:  File "/usr/lib/python3.8/site-packages/apex/optimizers/fused_adam.py", line 81, in __init__
[default3]:    super(FusedAdam, self).__init__(params, defaults)
[default6]:  File "/usr/lib/python3.8/site-packages/apex/optimizers/fused_adam.py", line 81, in __init__
[default6]:    super(FusedAdam, self).__init__(params, defaults)
[default6]:  File "/usr/lib/python3.8/site-packages/torch/optim/optimizer.py", line 61, in __init__
[default6]:    raise ValueError("optimizer got an empty parameter list")
[default3]:  File "/usr/lib/python3.8/site-packages/torch/optim/optimizer.py", line 61, in __init__
[default3]:    raise ValueError("optimizer got an empty parameter list")

`

**Comment by ygfrancois** [comment]
> @ygfrancois Can you show the error message?
> 
> In addition, I think tp=4 will distribute embedding and lmhead equally to each gpu. For example, when vocab size is 32000, theoretically the embedding size on each gpu is 32000/4=8000. Since alignment is 128 (controlled by `--make_vocab_size_divisible_by`), The actual embedding size will be 8064. That means each gpu should have trainable parameters.

And，i have a new question. 
If I want to warm up train only some special token in the SFT stage(I found the initialization of special token which have not been pretrained influence a lot to the SFT training). Which means maybe only 2 word embedding weights and 2 lm_head weights need to train.
In this case, there maybe also optimizer problem when using tp=8.

**Comment by jubgjf** [comment]
I've only tested the code with pp=1. This problem does occur when pp>1, but I haven't thought of a good solution yet. 

However, considering that you have frozen most of the model weights, you may try lowering the pp level to pp=1. If you use bf16 or fp16 for training, since most parameters are not involved in gradient calculation and optimizer state, I guess using tensor parallelism alone is enough for training the 70B model.

**Comment by ygfrancois** [comment]
> I've only tested the code with pp=1. This problem does occur when pp>1, but I haven't thought of a good solution yet.
> 
> However, considering that you have frozen most of the model weights, you may try lowering the pp level to pp=1. If you use bf16 or fp16 for training, since most parameters are not involved in gradient calculation and optimizer state, I guess using tensor parallelism alone is enough for training the 70B model.

thank you.

pp=1 is ok for me, but the sequence length could not be set to 8192 due to the GPT memory
